File: harness/bunnymark/src/main/kotlin/godot/benchmark/bunnymark/BunnymarkV1DrawTexture.kt
package godot.benchmark.bunnymark

import godot.Node2D
import godot.RandomNumberGenerator
import godot.ResourceLoader
import godot.Texture
import godot.annotation.RegisterClass
import godot.annotation.RegisterFunction
import godot.annotation.RegisterSignal
import godot.core.Vector2
import godot.signals.signal

@RegisterClass("BunnymarkV1DrawTexture")
class BunnymarkV1DrawTexture : Node2D() {

    @RegisterSignal
    val signalBenchmarkFinished by signal<Int>("bunnyCount")

    data class Bunny(var position: Vector2, var speed: Vector2)

    private val bunnies = mutableListOf<Bunny>()
    private val gravity = 500
    private val bunnyTexture = ResourceLoader.load("res://images/godot_bunny.png") as Texture
    private val randomNumberGenerator = RandomNumberGenerator()

    private lateinit var screenSize: Vector2

    @RegisterFunction
    override fun _ready() {
        randomNumberGenerator.randomize()
    }

    @RegisterFunction
    override fun _draw() {
        for (bunny in bunnies) {
            drawTexture(bunnyTexture, bunny.position)
        }
    }

    @RegisterFunction
    override fun _process(delta: Double) {
        screenSize = getViewportRect().size

        for (bunny in bunnies) {
            val pos = bunny.position
            val speed = bunny.speed

            pos.x += speed.x * delta
            pos.y += speed.y * delta

            speed.y += gravity * delta

            if (pos.x > screenSize.x) {
                speed.x *= -1
                pos.x = screenSize.x
            }

            if (pos.x < 0) {
                speed.x *= -1.0
                pos.x = 0.0
            }

            if (pos.y > screenSize.y) {
                pos.y = screenSize.y
                if (randomNumberGenerator.randf() > 0.5) {
                    speed.y = -(randomNumberGenerator.randi() % 1100 + 50).toDouble()
                } else {
                    speed.y *= -0.85
                }
            }

            if (pos.y < 0) {
                speed.y = 0.0
                pos.y = 0.0
            }

            bunny.position = pos
            bunny.speed = speed
        }
        update()
    }

    @RegisterFunction
    fun addBunny() {
        bunnies.add(
            Bunny(
                Vector2(screenSize.x / 2, screenSize.y / 2),
                Vector2(randomNumberGenerator.randi() % 200 + 50, randomNumberGenerator.randi() % 200 + 50)
            )
        )
    }

    @RegisterFunction
    fun removeBunny() {
        if (bunnies.size == 0) return
        bunnies.removeAt(bunnies.size - 1)
    }

    @RegisterFunction
    fun finish() {
        signalBenchmarkFinished.emit(bunnies.size)
    }
}

File: harness/bunnymark/benchmarks/BunnymarkV1DrawTexture/cs/BunnymarkV1DrawTexture.cs
using Godot;
using System;
using System.Collections.Generic;

public partial class BunnymarkV1DrawTexture : Node2D
{
    private class Pair
    {
        public Vector2 Current;
        public Vector2 Next;
    }

    List<Pair> bunnies = new List<Pair>();
    Vector2 screenSize;
    Texture2D bunnyTexture = (Texture2D)GD.Load("res://images/godot_bunny.png");
    Random random = new Random();
    int gravity = 500;

    public override void _Draw()
    {
        foreach (var bunny in bunnies)
            DrawTexture(bunnyTexture, bunny.Current);
    }
	
    public override void _Process(float delta)
    {
        screenSize = GetViewportRect().Size;

        foreach (var bunny in bunnies)
        {
            var position = bunny.Current;
            var newPosition = bunny.Next;

            position.x += newPosition.x * delta;
            position.y += newPosition.y * delta;

            newPosition.y += gravity * delta;

            if (position.x > screenSize.x)
            {
                newPosition.x *= -1;
                position.x = screenSize.x;
            }

            if (position.x < 0)
            {
                newPosition.x *= -1;
                position.x = 0;
            }

            if (position.y > screenSize.y)
            {
                position.y = screenSize.y;
                if (random.NextDouble() > 0.5)
                {
                    newPosition.y = (random.Next() % 1100 + 50);
                }
                else
                {
                    newPosition.y *= -0.85f;
                }
            }

            if (position.y < 0)
            {
                newPosition.y = 0;
                position.y = 0;
            }
			
            bunny.Current = position;
            bunny.Next = newPosition;
        }

        Update();
    }

    public void add_bunny()
    {
        bunnies.Add(new Pair() { Current = new Vector2(screenSize.x / 2, screenSize.y / 2), Next = new Vector2(random.Next() % 200 + 50, random.Next() % 200 + 50) });
    }

    public void remove_bunny()
    {
        if (bunnies.Count == 0) {
            return;
        }
		
        bunnies.RemoveAt(bunnies.Count - 1);
    }

    public void finish()
    {
        EmitSignal("benchmark_finished", bunnies.Count);
    }
}
